### PR TITLE
cv-with-pre-constructed-datasets

### PR DESCRIPTION
**Pull request by chilledgeek**
Modified functions cv and search_cv to allow inputs with pre-constructed datasets.

This will speed up the process if it is part of an iterative process and a pre-constructed dataset is available

### CODE REVIEW

File: test/basic/cv.jl [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    # Assert
+    @test output isa Dict{String,Dict{String,Vector{Float64}}}
+    @test length(output["validation"]["auc"]) == 5
+    @test length(output["training"]["auc"]) == 5
+end
+
+
+
+end # module
```
Comment (yalwan-iqvia) [inline test/basic/cv.jl:51]: No newline at end of file -- need to change a setting in your editor perhaps?

File: test/basic/cv.jl [hunk 2]
```diff
@@ -0,0 +1,51 @@
+module TestFit
```
Comment (yalwan-iqvia) [inline test/basic/cv.jl:1]: Careful here, you're effectively overwriting a modeule (it probably generates a warning during test runs)

File: test/basic/search_cv.jl [hunk 3]
```diff
@@ -0,0 +1,60 @@
+module TestFit
+
+include("../../src/LightGBM.jl")
+
+using Test
+using LightGBM
+
+train_matrix = rand(1000,70) # create random dataset
```
Comment (yalwan-iqvia) [inline test/basic/search_cv.jl:8]: Nice, thanks for adding tests for this functionality

File: src/cv.jl [hunk 4]
```diff
@@ -15,30 +17,38 @@ last valid iteration.
...
 
+    return cv(estimator, full_ds, splits, verbosity = verbosity)
+end
+
+# Pass Dataset class directly. This will speed up the process if it is part of an iterative process and a pre-constructed dataset is available
+function cv(estimator::LGBMEstimator, dataset::Dataset, splits;
+    verbosity::Integer = 1)
+    start_time = now()
+    num_data = LGBM_DatasetGetNumData(dataset)
+    ds_parameters = stringifyparams(estimator, DATASETPARAMS)
```
Comment (yalwan-iqvia) [inline src/cv.jl:39]: Can you make it so verbosity is added on this line too?

File: src/cv.jl [hunk 5]
```diff
@@ -15,30 +17,38 @@ last valid iteration.
...
-
     full_ds = LGBM_DatasetCreateFromMat(X, ds_parameters)
     LGBM_DatasetSetField(full_ds, "label", y)
 
+    return cv(estimator, full_ds, splits, verbosity = verbosity)
+end
+
+# Pass Dataset class directly. This will speed up the process if it is part of an iterative process and a pre-constructed dataset is available
+function cv(estimator::LGBMEstimator, dataset::Dataset, splits;
+    verbosity::Integer = 1)
```
Comment (yalwan-iqvia) [inline src/cv.jl:36]: I think this doesn't need to be split across lines

File: test/basic/cv.jl (hunk 2 above)
Comment (chilledgeek) [reply test/basic/cv.jl:1]: o yes thanks for spotting that! will also modify that for the `TestSearchCv` as well in my next push

**Review by yalwan-iqvia — APPROVED** [review]

**Pull request merged**